pkg/csi/service/wcpguest/controller.go

```diff
@@ -307,6 +326,121 @@ func (c *controller) ControllerPublishVolume(ctx context.Context, req *csi.Contr
+			// When deletionTimestamp is set, CnsOperator is in the process of removing access for this IP.
+			// When that operation is successful, the instance will be deleted. In a subsequent retry, a new instance will be created.
+			msg := fmt.Sprintf("failed to get VirtualMachines for the node: %q. Error: %+v", req.NodeId, err)
+			log.Error(msg)
+			return nil, status.Errorf(codes.Internal, msg)
+		}
+		// Verify if the CnsFileAccessConfig instance has status with done set to true and error is empty
+		if cnsFileAccessConfigInstance.Status.Done && cnsFileAccessConfigInstance.Status.Error == "" {
+			resp := &csi.ControllerPublishVolumeResponse{
+				PublishContext: cnsFileAccessConfigInstance.Status.AccessPoints,
+			}
+			log.Infof("ControllerPublishVolume: Volume attached successfully %q", req.VolumeId)
+			return resp, nil
+		}
+		cfg, err := common.GetConfig(ctx)
+		if err != nil {
+			msg := fmt.Sprintf("failed to read config. Error: %+v", err)
+			log.Error(msg)
+			return nil, status.Errorf(codes.Internal, msg)
+		}
+		restClientConfig := k8s.GetRestClientConfigForSupervisor(ctx, cfg.GC.Endpoint, cfg.GC.Port)
+		cnsFileAccessConfigWatcher, err := k8s.NewCnsFileAccessConfigWatcher(ctx, restClientConfig, c.supervisorNamespace)
+		if err != nil {
+			msg := fmt.Sprintf("failed to create cnsFileAccessConfigWatcher. Error: %+v", err)
+			log.Error(msg)
+			return nil, status.Errorf(codes.Internal, msg)
+		}
+		// Attacher timeout, default is set to 4 minutes
+		timeoutSeconds := int64(getAttacherTimeoutInMin(ctx) * 60)
+		// Adding watch on the CnsFileAccessConfig instance to register for updates
+		watchCnsFileAccessConfig, err := cnsFileAccessConfigWatcher.Watch(metav1.ListOptions{
+			FieldSelector:   fields.SelectorFromSet(fields.Set{"metadata.name": cnsFileAccessConfigInstance.Name}).String(),
+			ResourceVersion: cnsFileAccessConfigInstance.ResourceVersion,
+			TimeoutSeconds:  &timeoutSeconds,
+		})
+		if err != nil {
+			msg := fmt.Sprintf("failed to watch virtualMachine %q with Error: %v", virtualMachine.Name, err)
+			log.Error(msg)
+			return nil, status.Errorf(codes.Internal, msg)
+		}
+		defer watchCnsFileAccessConfig.Stop()
+		timeout := time.Now().Add(time.Duration(timeoutSeconds) * time.Second)
+		// Watch all update events made on CnsFileAccessConfig instance until accessPoints is set
+		for {
+			log.Debugf("Waiting for update on cnsfileaccessconfigs: %q", cnsFileAccessConfigInstance.Name)
+			event := <-watchCnsFileAccessConfig.ResultChan()
+			cnsfileaccessconfig, ok := event.Object.(*cnsfileaccessconfigv1alpha1.CnsFileAccessConfig)
+			if !ok {
+				msg := fmt.Sprintf("Watch on cnsfileaccessconfig instance %q timed out", cnsFileAccessConfigInstance.Name)
+				log.Error(msg)
+				return nil, status.Errorf(codes.Internal, msg)
+			}
+			log.Debugf("Observed an update on cnsfileaccessconfig: %+v", cnsfileaccessconfig)
+			if cnsfileaccessconfig.Status.Done && cnsfileaccessconfig.Status.Error == "" {
```

Comment:
Dont we have to check the cnsfileaccessconfig CR is for the VM and SV PVC instance? Apparently the watcher returns other cnsfileaccessconfig CRs too. We had similar problem for VirtualMachine CR and we handled it by checking the name (see line 519 in this file).

Reply:
Done. Thanks for suggesting this. I was not aware of this bug.